### PR TITLE
Support additional units for INTERVAL types #11490

### PR DESCRIPTION
**Pull request by marstein**
Added milliseconds parsing and arithmetic analogue to the seconds. Added tests following the examples from second intervals.

## Summary of the changes / Why this improves CrateDB
Added parsing of milliseconds values to interval literals and expressions.
Some customers request to specify milliseconds in `INTERVAL` expressions.

It is now possible to specify intervals with milliseconds and work with them: `select '3 days 400 milliseconds`::interval;` etc

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
ok to test

File: docs/appendices/release-notes/5.9.0.rst [hunk 1]
```diff
@@ -75,7 +75,7 @@ Changes
 SQL Statements
 --------------
 
-None
+The ``INTERVAL`` data type now supports milliseconds. Intervals can be specified with msec, ms, msecs, millisecond or milliseconds, and ``MILLISECOND`` can be used in INTERVAL literals.
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/5.9.0.rst:78]: We mention external contributors in the release notes. 
Also, I think we should move it to the `SQL Standard and PostgreSQL Compatibility` section.

File: server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java [hunk 2]
```diff
@@ -62,6 +62,11 @@ public void testEvaluateInterval() {
         assertEvaluate("to_char(INTERVAL '1 year 2 months 3 weeks 5 hours 6 minutes 7 seconds', 'YYYY MM DD HH12:MI:SS')", "0001 03 22 05:06:07");
     }
 
+    @Test
+    public void testEvaluateIntervalMilliseconds() {
+        assertEvaluate("to_char(INTERVAL '1 year 2 months 3 weeks 5 hours 6 minutes 7 seconds 1000 milliseconds', 'YYYY MM DD HH12:MI:SS')", "0001 03 22 05:06:08");
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java:67]: Please add a test (can be another `assertEvaluate` call in the same test) with fractional milliseconds.

Here we add 7 seconds and 1000 milliseconds and get 8. 
Would be interesting to test smth like `... 7 seconds 500 milliseconds`

File: server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java [hunk 3]
```diff
@@ -62,6 +62,11 @@ public void testEvaluateInterval() {
         assertEvaluate("to_char(INTERVAL '1 year 2 months 3 weeks 5 hours 6 minutes 7 seconds', 'YYYY MM DD HH12:MI:SS')", "0001 03 22 05:06:07");
     }
 
+    @Test
+    public void testEvaluateIntervalMilliseconds() {
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java:66]: very minor: camelCase is used only in old tests. 
We use underscore (`test_evaluate_interval_milliseconds`) in new tests. Same applies for other camelCase instances

File: server/src/test/java/io/crate/interval/IntervalParserTest.java [hunk 4]
```diff
@@ -246,11 +260,17 @@ public void test_invalid_duplicate_units() {
         assertThatThrownBy(() -> IntervalParser.apply("1 years 2 mons 3 days 2 years"))
             .isExactlyInstanceOf(IllegalArgumentException.class)
             .hasMessage("Invalid interval format: 1 years 2 mons 3 days 2 years");
+        assertThatThrownBy(() -> IntervalParser.apply("5 millis 1 years 2 mons 3 days 2 ms"))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Invalid interval format: 5 millis 1 years 2 mons 3 days 2 ms");
+        assertThatThrownBy(() -> IntervalParser.apply("1 years 2 mons 3 days 2.2 secs 2 ms"))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Invalid interval format: 1 years 2 mons 3 days 2.2 secs 2 ms");
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/interval/IntervalParserTest.java:268]: Second assertion doesn't correlate with the test name `test_invalid_duplicate_units`. 
Let's remove it.  

1 years 2 mons 3 days 2.2 secs 2 ms" actually should be fine. 
I haven't checked locally, but unexpected `IllegalArgumentException` here seems to be related to `parseInteger` usage instead of `parseMilliSeconds(valueToken)` in `PGIntervalParser` (see also comment above)

**Review by BaurzhanSakhariev — COMMENTED** [review]
Hi @marstein thanks a lot for your contribution!

Looks good overall, added some comments

**Comment by BaurzhanSakhariev** [comment]
CI failure: `test_pg_get_keywords`
> Expected size: 279 but was: 280 

Please adjust an assertion in the test

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java [hunk 5]
```diff
@@ -112,13 +126,34 @@ public void testIntervalInvalidStartEnd() throws Exception {
...
+    public void test_odd_milliseconds() throws Exception {
+        var symbol = e.asSymbol("INTERVAL '101' MILLISECOND");
+        assertThat(symbol).isLiteral(new Period().withMillis(101)
+                                         .withPeriodType(PeriodType.yearMonthDayTime()));
+    }
+
+    @Test
+    public void test_odd_milliseconds_to_Second() throws Exception {
+        var symbol = e.asSymbol("INTERVAL '1.99' SECOND TO MILLISECOND");
+        assertThat(symbol).isLiteral(new Period().withSeconds(1).withMillis(990)
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:153]: naming: other way around: should be test_second_to_millisecond.

File: server/src/main/java/io/crate/interval/IntervalParser.java [hunk 6]
```diff
@@ -135,6 +139,9 @@ static Period roundToPrecision(Period period, Precision start, Precision end) {
             if (end == Precision.SECOND) {
                 return period.withMillis(0);
             }
+            if (end == Precision.MILLISECOND) {
+                return period.withMillis(0);
+            }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/interval/IntervalParser.java:144]: We should reset next unit to 0, not the `end` one. 
`UNIT1 TO UNIT2` means that we keep precision up to `UNIT2`


Since millis is the smallest unit, we should keep millis -> we should do nothing here and revert changes in the `IntervalParser. roundToPrecision` method. 

Technically speaking, we should replace 
```
if (end == Precision.MILLISECOND) {
    return period.withMillis(0);
}
```

with smth like
```
if (end == Precision.MILLISECOND) {
    return period.withMicros(0);
}
```

But Period doesn't have micros granularity, millis is already the smallest unit so we have nothing to reset.

BTW, not related to this PR but just loudly thinking:

A consequence of Period not having mircoseconds is that we won't be able to do `UNIT TO MILLISECONDS` when when we eventually support `MICROSECOND` unit for INTERVAL - we would have to reset microseconds but won't be able to do so - but now all good as we don't support it.

 `Period` doesn't have `microsecond` unit, so , we would have to reset

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/interval/IntervalParser.java:144]: We need more tests with `SOME_UNIT TO MILLISECONDS` in the `IntervalParserTest` to cover this. 

`IntervalAnalysisTest` tests don't cover it, it doesn't call IntervalParser methods.

File: server/src/main/java/io/crate/interval/NumericalIntervalParser.java [hunk 7]
```diff
@@ -97,8 +97,16 @@ private static Period roundToPrecision(int value,
...
+            if (end == null) {
+                return buildSecondsWithMillisPeriod(value, millis);
+            }
+            if (end == IntervalParser.Precision.MILLISECOND) {
+                return buildSecondsWithMillisPeriod(0, value * 1000 + millis);
+            }
+        }
+        if (start == IntervalParser.Precision.MILLISECOND && end == null) {
+            return buildSecondsWithMillisPeriod(0, value);
         }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/interval/NumericalIntervalParser.java:110]: Please add `IntervalParserTest` for this

File: server/src/main/java/io/crate/interval/NumericalIntervalParser.java [hunk 8]
```diff
@@ -97,8 +97,16 @@ private static Period roundToPrecision(int value,
...
+        if (start == IntervalParser.Precision.SECOND) {
+            if (end == null) {
+                return buildSecondsWithMillisPeriod(value, millis);
+            }
+            if (end == IntervalParser.Precision.MILLISECOND) {
+                return buildSecondsWithMillisPeriod(0, value * 1000 + millis);
+            }
+        }
+        if (start == IntervalParser.Precision.MILLISECOND && end == null) {
+            return buildSecondsWithMillisPeriod(0, value);
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/interval/NumericalIntervalParser.java:109]: Is it possible scenario to have MILLISECOND precision as start?  
See also `testIntervalInvalidStartEndMillis`

File: server/src/main/java/io/crate/interval/NumericalIntervalParser.java [hunk 9]
```diff
@@ -97,8 +97,16 @@ private static Period roundToPrecision(int value,
...
-        if (start == IntervalParser.Precision.SECOND && end == null) {
-            return buildSecondsWithMillisPeriod(value, millis);
+        if (start == IntervalParser.Precision.SECOND) {
+            if (end == null) {
+                return buildSecondsWithMillisPeriod(value, millis);
+            }
+            if (end == IntervalParser.Precision.MILLISECOND) {
+                return buildSecondsWithMillisPeriod(0, value * 1000 + millis);
+            }
+        }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/interval/NumericalIntervalParser.java:107]: I think we can revert NumericalIntervalParser changes. 

If start is SECONDS, we should never truncate it, ie shouldn't do `buildSecondsWithMillisPeriod(0...`

Pretty much same logic as in another comment - millis it the smallest unit for now, we cannot start with it and we have nothing below to reset.

File: server/src/main/java/io/crate/interval/IntervalParser.java (hunk 6 above)
Comment (marstein) [reply server/src/main/java/io/crate/interval/IntervalParser.java:144]: When checking in the [Postgres manual](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT:~:text=The%20interval%20type%20has%20an%20additional%20option%2C%20which%20is%20to%20restrict%20the%20set%20of%20stored%20fields%20by%20writing%20one%20of%20these%20phrases%3A), it says 

> The interval type has an additional option, which is to restrict the set of stored fields by writing one of these phrases
```
YEAR
MONTH
DAY
HOUR
MINUTE
SECOND
YEAR TO MONTH
DAY TO HOUR
DAY TO MINUTE
DAY TO SECOND
HOUR TO MINUTE
HOUR TO SECOND
MINUTE TO SECOND
```

So probably MILLISECONDS should not be in there at all.

Also, if I test the behavior in Postgres, the second fractions behave differently, like you allueded to above:

```
# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' HOUR TO SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4

teslamate=# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 400 ms' MINUTE TO SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4
(1 row)

teslamate=# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 400 ms' SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4
(1 row)
```

Basically seconds are treated as floating point numbers, and ms are not cut off.


Should I change the implementation to reflect this?

File: server/src/main/java/io/crate/interval/NumericalIntervalParser.java (hunk 9 above)
Comment (marstein) [reply server/src/main/java/io/crate/interval/NumericalIntervalParser.java:107]: Agree. After checking in Postgres, the seconds part is treated as one unit. The milliseconds are port of the seconds.

File: server/src/main/java/io/crate/interval/IntervalParser.java (hunk 6 above)
Comment (marstein) [reply server/src/main/java/io/crate/interval/IntervalParser.java:144]: For completeness here is what crate outputs:
```
cr> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' HOUR TO SECOND;
+--------------------------------+
| 'P4Y3M2DT1H2M4S'::interval     |
+--------------------------------+
| 4 years 3 mons 2 days 01:02:04 |
+--------------------------------+
SELECT 1 row in set (0.014 sec)
cr> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' DAY TO SECOND;
+--------------------------------+
| 'P4Y3M2DT1H2M4S'::interval     |
+--------------------------------+
| 4 years 3 mons 2 days 01:02:04 |
+--------------------------------+
```

The milliseconds are unset/zero.

File: server/src/main/java/io/crate/interval/PGIntervalParser.java [hunk 10]
```diff
@@ -168,6 +168,13 @@ static Period apply(String value) {
                             milliSeconds = parseMilliSeconds(valueToken);
                             timeParsed = true;
                         }
+                        case "ms", "msec", "msecs", "millisecond", "milliseconds" -> {
+                            if (milliSeconds > 0) {
+                                throw new IllegalArgumentException("Invalid interval format: " + value);
+                            }
+                            milliSeconds = parseInteger(valueToken);
```
[the remark this answers is not in the document]
Comment (marstein) [reply server/src/main/java/io/crate/interval/PGIntervalParser.java:175]: Okay, I used parseMilliSeconds.

After looking at the Postgres docs and trying it out in Postgres, I agree that seconds and millis should be seen as one. ie. the zeroing out of the millis should not be done. I followed the precedent here, but I think when the end-precision is SECOND then the millis should be preserved. I shall change things to do that. Please also see my comment below.

File: docs/appendices/release-notes/5.9.0.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (marstein) [reply docs/appendices/release-notes/5.9.0.rst:78]: Thanks. Done.

File: server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java (hunk 2 above)
Comment (marstein) [reply server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java:67]: Done. Added several. 
It turns out that in `io.crate.interval.PGIntervalParser#apply(java.lang.String)` line 177 it is necessary to use `parseInteger()`, otherwise the int portion of the seconds (the 1 in 1.7) would be cut off.

File: server/src/test/java/io/crate/interval/IntervalParserTest.java (hunk 4 above)
Comment (marstein) [reply server/src/test/java/io/crate/interval/IntervalParserTest.java:268]: Once 2.2 seconds is specified, 200 ms are implied in the underlying Period instance, because the ms are treated in Postgres like part of the seconds. ie 2.2s evaluates to a Period with 2s and 200ms, and then ms are set and another mention of ms would trigger the invalid argument exception.

Postgres handles it the same:
```
teslamate=# select '1 year 2 months 3 weeks 5 hours 6 minutes 7.7 seconds 1000 milliseconds'::interval;
ERROR:  invalid input syntax for type interval: "1 year 2 months 3 weeks 5 hours 6 minutes 7.7 seconds 1000 milliseconds"
LINE 1: select '1 year 2 months 3 weeks 5 hours 6 minutes 7.7 second...
               ^
teslamate=# select '7 seconds 100 milliseconds'::interval;
  interval
------------
 00:00:07.1
(1 row)

teslamate=# select '7.1 seconds 100 milliseconds'::interval;
ERROR:  invalid input syntax for type interval: "7.1 seconds 100 milliseconds"
LINE 1: select '7.1 seconds 100 milliseconds'::interval;
               ^
```

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java (hunk 5 above)
Comment (marstein) [reply server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:153]: I removed the `SECOND TO MILLISECOND` limiter because it is not in Postgres.

File: server/src/main/java/io/crate/interval/NumericalIntervalParser.java (hunk 8 above)
Comment (marstein) [reply server/src/main/java/io/crate/interval/NumericalIntervalParser.java:109]: The interval in Postgres does not support MILLISECOND as start or end, so I removed it.

File: docs/general/ddl/data-types.rst [hunk 11]
```diff
@@ -1987,7 +1988,7 @@ Where ``string`` describes the interval using one of the recognized formats:
 | Standard SQL format  | ``3 4:05:06``         | 3 days 4 hours      |
 | (day-time)           |                       | 5 minutes 6 seconds |
 +----------------------+-----------------------+---------------------+
-| `PostgreSQL interval | ``1 year 2 months     | 1 year 2 months     |
+| `PostgreSQL interval | ``1 year 2 months``   | 1 year 2 months     |
```
Comment (matriv) [inline docs/general/ddl/data-types.rst:1991]: This should be reverted, the backticks are closing at the end of the expression (6 seconds).

File: server/src/test/java/io/crate/interval/IntervalParserTest.java (hunk 4 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/interval/IntervalParserTest.java:268]: Yes, fractional sec + ms == twice ms, thanks for checking.

Maybe worth adding a comment for the second assertion as it looks confusing/not obvious.

File: server/src/main/java/io/crate/interval/IntervalParser.java (hunk 6 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/interval/IntervalParser.java]: I thought millis are kept only when they are part of seconds (fractional part of X sec) but even integer sec integer ms keeps millis in Postgres 15:

> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 2 ms' DAY TO SECOND;
> {"years":4,"months":3,"days":2,"hours":1,"minutes":2,"seconds":4,"milliseconds":2}

So, regarding `TO MILLISECONDS`, I agree that it shouldn't be even supported.

And regarding CrateDB truncating milliseconds in TO SECOND - I would rather create a dedicated issue for that in order to keep the scope of this PR small. We don't always behave 1:1 with PG, so would also need to dicsuss wheter we treat it as a bug, or do it as a feature or don't do at all - all there is not really relevant for MILLISECOND unit task.

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java [hunk 12]
```diff
@@ -44,37 +44,65 @@ public void prepare() throws Exception {
...
+        assertThat(symbol).isLiteral(new Period().withYears(6).withMonths(5).withDays(4).withHours(3).withMinutes(2)
+            .withSeconds(0)
+            .withPeriodType(PeriodType.yearMonthDayTime()));
+    }
+
+    @Test
+    public void test_psql_compact_format_from_string_with_start_end_hour_seconds() {
+        var symbol = e.asSymbol("INTERVAL '6 years 5 mons 4 days 03:02:01' HOUR TO SECOND");
+        assertThat(symbol).isLiteral(new Period().withYears(6).withMonths(5).withDays(4).withHours(3).withMinutes(2).withSeconds(1)
+            .withPeriodType(PeriodType.yearMonthDayTime()));
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:105]: This is not related to MILLISECOND change, I think we can remove it. 

On the other hand, it's better coverage of the existing functionality, isn't it? 
If you want, we can keep it but let's put those changes to a separate commit on order not to mix it with MILLISECONDS unit changes.

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java [hunk 13]
```diff
@@ -143,14 +164,8 @@ public void test_odd() throws Exception {
...
-        assertThat(symbol).isLiteral(new Period().withMillis(101)
-                                         .withPeriodType(PeriodType.yearMonthDayTime()));
-    }
-
-    @Test
-    public void test_odd_milliseconds_to_Second() throws Exception {
-        var symbol = e.asSymbol("INTERVAL '1.99' SECOND TO MILLISECOND");
-        assertThat(symbol).isLiteral(new Period().withSeconds(1).withMillis(990)
+        // @todo this is strange, but Postgres says so too: # select INTERVAL '101' MILLISECOND; ->  00:01:41
+        assertThat(symbol).isLiteral(new Period().withMinutes(1).withSeconds(41)
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:168]: Because it's synonym of `select INTERVAL '101' as MILLISECOND` ie it's a column name.

To get 101 millis, we should put keyword inside quotes: `select INTERVAL '101 MILLISECOND' ;`

File: server/src/main/java/io/crate/interval/PGIntervalParser.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/interval/PGIntervalParser.java]: >  I shall change things to do that

Yes, agree. I think both `test_psql_compact_format_from_string_with_milliseconds` and `test_psql_compact_format_from_string_with_milliseconds_from_minute` which are removed now are actually useful - but should return `.withMillis(200)`

File: server/src/main/java/io/crate/interval/IntervalParser.java (hunk 6 above)
Comment (marstein) [reply server/src/main/java/io/crate/interval/IntervalParser.java]: @BaurzhanSakhariev for the last item I created <a href='https://github.com/crate/crate/issues/16338'>INTERVAL <interval> <UNIT> TO SECOND expressions truncate milliseconds, unlike Postgres · Issue #16338 · crate/crate</a> and removed the code implementing the change to leaving the ms intact.

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java (hunk 12 above)
Comment (marstein) [reply server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:105]: Thanks. I will make a separate commit for these tests.

File: server/src/test/java/io/crate/interval/IntervalParserTest.java (hunk 4 above)
Comment (marstein) [reply server/src/test/java/io/crate/interval/IntervalParserTest.java:268]: ✅ Added comment on the assertion.

File: server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java (hunk 13 above)
Comment (marstein) [reply server/src/test/java/io/crate/analyze/expressions/IntervalAnalysisTest.java:168]: Got it - I added another assert so that both cases are illustrated.

File: server/src/main/java/io/crate/interval/PGIntervalParser.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (marstein) [reply server/src/main/java/io/crate/interval/PGIntervalParser.java]: I added them back, but `.withMillis(0)` as you suggested in the other comment and filed issue #16338

File: docs/general/ddl/data-types.rst (hunk 11 above)
Comment (marstein) [reply docs/general/ddl/data-types.rst:1991]: Done. Thanks.

File: server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java (hunk 3 above)
Comment (marstein) [reply server/src/test/java/io/crate/expression/scalar/formatting/ToCharFunctionTest.java]: Fixed.

File: server/src/main/java/io/crate/interval/IntervalParser.java (hunk 6 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/interval/IntervalParser.java]: Thanks!

File: server/src/main/java/io/crate/interval/PGIntervalParser.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (marstein) [reply server/src/main/java/io/crate/interval/PGIntervalParser.java]: Thanks for the clarification.

For 1. things work as expected (`io.crate.interval.PGIntervalParser#apply(java.lang.String)` does the right thing - millis are preserved correctly. I added explicit tests for that in `io.crate.analyze.expressions.IntervalAnalysisTest#test_seconds_millis` to check that intervals with millisecond values come out as expected:

```java
        symbol = e.asSymbol("INTERVAL '1000 milliseconds'");
        assertThat(symbol).isLiteral(new Period().withSeconds(1) ...

        symbol = e.asSymbol("INTERVAL '1 secs 100 ms'");
        assertThat(symbol).isLiteral(new Period().withSeconds(1).withMillis(100) ...

        symbol = e.asSymbol("INTERVAL '60 secs 100 ms'");
        assertThat(symbol).isLiteral(new Period().withMinutes(1).withMillis(100) ...
```

For 2. I dropped the `TO SECOND` and made sure that seconds/milliseconds come across right with unit focus and without.

**Comment by marstein** [comment]
@BaurzhanSakhariev @matriv I think everything is ready now.

**Comment by BaurzhanSakhariev** [comment]
> I think everything is ready now.

Thanks, I skimmed through and looks good. 

I want to test couple more things locally and then I will merge it - but not today.

**Review by BaurzhanSakhariev — APPROVED** [review]

**Comment by BaurzhanSakhariev** [comment]
Hi @marstein, thanks a lot for your contribution and for patience with all comments!

I did some cosmetic changes (removed code comments, fixed link in the change entry, removed one more unrelated change in the IntervalParser, updated commit messages) - but the core changes are left intact.
